modules/mod-deep-learning/DeepModelManager.cpp

```diff
@@ -0,0 +1,554 @@
+/**********************************************************************
+
+   Audacity: A Digital Audio Editor
+   Audacity(R) is copyright (c) 1999-2021 Audacity Team.
+
+   DeepModelManager.cpp
+   Hugo Flores Garcia
+
+******************************************************************/
+
+#include "DeepModelManager.h"
+
+#include "CodeConversions.h"
+#include "NetworkManager.h"
+#include "IResponse.h"
+#include "Request.h"
+
+#include <wx/tokenzr.h>
+#include <wx/log.h>
+
+using namespace audacity;
+ 
+DeepModelManager::DeepModelManager() :
+   mAPIEndpoint("https://huggingface.co/api/")
+{
+   const std::string schemaPath = audacity::ToUTF8(
+      wxFileName(DeepModel::BuiltInModulesDir(), wxT("modelcard-schema.json"))
+                  .GetFullPath()
+   );
+   mModelCardSchema = parsers::ParseFile(schemaPath);
+}
+
+DeepModelManager& DeepModelManager::Get()
+{
+   static DeepModelManager manager;
+   return manager;
+}
+
+FilePath DeepModelManager::GetRepoDir(ModelCardHolder card) const
+{
+   static const std::string sep = "_";
+
+   FilePath repoDir = FileNames::MkDir( 
+      wxFileName(DeepModel::DLModelsDir(),
+                 card->author() + sep + card->name() 
+      ).GetFullPath()
+   );
+
+   return repoDir;
+}
+
+DeepModelHolder DeepModelManager::GetModel(ModelCardHolder card) const
+{
+   if (!IsInstalled(card))
+      throw ModelManagerException(XO("Model is not loaded."));
+
+   DeepModelHolder model = std::make_shared<DeepModel>();
+   model->SetCard(card);
+
+   // GetRepoDir won't work if the card is empty
+   wxFileName path = wxFileName(GetRepoDir(card), "model.pt");
+
+   // finally, load
+   model->Load(audacity::ToUTF8(path.GetFullPath()));
+
+   return model;
+}
+
+std::string DeepModelManager::GetMoreInfoURL(ModelCardHolder card) const
+{
+   return "https://huggingface.co/" + card->GetRepoID();
+}
+
+bool DeepModelManager::IsInstalled(ModelCardHolder card) const
+{
+   FilePath repoDir = wxFileName(card->local_path()).GetFullPath();
+
+   wxFileName modelPath = wxFileName(repoDir, "model.pt");
+   wxFileName metadataPath = wxFileName(repoDir, "metadata.json");
+
+   return (modelPath.FileExists() && metadataPath.FileExists());
+}
+
+void DeepModelManager::Install(ModelCardHolder card, ProgressCallback onProgress, CompletionHandler onCompleted)
+{ 
+   if (IsInstalled(card))
+      return ;
+
+   ProgressCallback progressHandler(
+   [this, card, handler {std::move(onProgress)}](int64_t current, int64_t expected)
+   {
+      // if the install has been cancelled, bail
+      // if we don't bail early, then calling the handler will 
+      // segfault.
+      if (!this->IsInstalling(card))
+         return;
+
+      return handler(current, expected);
+   }
+   );
+
+   // set up the install handler
+   CompletionHandler installHandler(
+   [this, card, handler {std::move(onCompleted)} ](int httpCode, std::string body)
+   {
+      if (!this->IsInstalling(card))
+      {
+         Uninstall(card);
```

Comment:
`Uninstall` throws if the card was already not installed.  Are you sure it won't throw here?  Maybe make Uninstall non-throwing, and insert throws at other uses of it?